openassessment/xblock/static/js/src/lms/oa_response.js

```diff
@@ -474,71 +537,259 @@ OpenAssessment.ResponseView.prototype = {
+                this.baseView.toggleActionError(
+                    'upload',
+                    gettext("File type is not allowed.")
+                );
+                errorCheckerTriggered = true;
+                break;
+            }
+        }
+
+        if (!errorCheckerTriggered) {
             this.baseView.toggleActionError('upload', null);
             this.files = files;
+            this.updateFilesDescriptionsFields(files, descriptions, uploadType);
+        }
+
+        if (this.files === null) {
+            sel.find('.file__upload').prop('disabled', true);
+        }
+    },
+
+    /**
+     Render textarea fields to input description for each uploaded file.
+
+     */
+    /* jshint -W083 */
+    updateFilesDescriptionsFields: function(files, descriptions, uploadType) {
+        var filesDescriptions = $(this.element).find('.files__descriptions').first();
+        var mainDiv = null;
+        var divLabel = null;
+        var divTextarea = null;
+        var divImage = null;
+        var img = null;
+        var textarea = null;
+        var descriptionsExists = true;
+
+        this.filesDescriptions = descriptions || [];
+
+        $(filesDescriptions).show().html('');
+
+        for (var i = 0; i < files.length; i++) {
+            mainDiv = $('<div/>');
+
+            divLabel = $('<div/>');
+            divLabel.addClass('submission__file__description__label');
+            divLabel.text(gettext("Describe ") + files[i].name + ' ' + gettext("(required):"));
+            divLabel.appendTo(mainDiv);
+
+            divTextarea = $('<div/>');
+            divTextarea.addClass('submission__file__description');
+            textarea = $('<textarea />');
+            if ((this.filesDescriptions.indexOf(i) !== -1) && (this.filesDescriptions[i] !== '')) {
+                textarea.val(this.filesDescriptions[i]);
+            } else {
+                descriptionsExists = false;
+            }
+            textarea.addClass('file__description file__description__' + i);
+            textarea.appendTo(divTextarea);
+
+            if (uploadType === "image") {
+                img = $('<img/>', {
+                    src: window.URL.createObjectURL(files[i]),
+                    height: 80
+                });
+                img.onload = function() {
+                    window.URL.revokeObjectURL(this.src);
+                };
+
+                divImage = $('<div/>');
+                divImage.addClass('submission__img__preview');
+                img.appendTo(divImage);
+                divImage.appendTo(mainDiv);
+            }
+
+            divTextarea.appendTo(mainDiv);
+
+            mainDiv.appendTo(filesDescriptions);
+            textarea.on("change keyup drop paste", $.proxy(this, "checkFilesDescriptions"));
         }
-        $(".file__upload").prop('disabled', this.files === null);
+
+        $(this.element).find('.file__upload').prop('disabled', !descriptionsExists);
```

Comment:
@efischer19 
There could be two different situations.

**1**. If in the ORA settings `Text Response == optional` and `File Uploads Response == required` in this case Submit button will be locked until user will upload at least one file.

**2**. If in the ORA settings `Text Response == optional` and also `File Uploads Response == optional` (I may suggest that your described case above is this one) - in this case Submit button will be unlocked if user will type something in the "response" textarea or if at least one file will be uploaded. File descriptions are required only for "Upload" button. So you may type anything in the textarea field, choose files, decide not to upload anything and submit the answer without upload.

Do you suggest to lock "Submit" button even if `File Uploads Response == optional` but files are chosen but not uploaded?